### PR TITLE
fixes #351, replaces none critical throws with collector.addError

### LINKED ISSUE #351: All errors unless totally fatal should add errors to the collector not throw.
This bug is to review that is the case and if not fix it.

Blocks #348

### PR DESCRIPTION
**Pull request by hameleonka**
I replaced some throws with collector.addError, just not sure if I got it right, If you(@EnTeQuAk) ok with this approach, I will finish this job. Working on this problem as an applicant for outreach program Dec 2017 -March 2018. Thanks.

### CODE REVIEW

**Comment by EnTeQuAk** [comment]
Hey @hameleonka, thank you so much for that pull request! It looks generally quite good and I don't see anything being majorly wrong so please feel free to go ahead and add a few tests if they're still needed. I'll have a detailed look and review later today, I'm so sorry for the late response.

**Comment by hameleonka** [comment]
you are welcome, waiting for your review, I think this PR covers all the functionality described in the issue, there are some throws left, but in my opinion they are critical.

File: docs/rules.md [hunk 1]
```diff
@@ -187,6 +187,12 @@ TODO: A lot of these are generated so this will need expanded with each unique c
 | :white_check_mark: | error | | | top level guid missing from install.rdf | install.rdf | | | RDF_ID_MISSING |
 | :white_check_mark: | error | | | top level version missing from install.rdf. | install.rdf | | | RDF_VERSION_MISSING |
 | :white_check_mark: | error | | | top level name missing from install.rdf | install.rdf | | | RDF_NAME_MISSING |
+| :white_check_mark: | error | | | RDF tag should have only one child  | install.rdf | | | RDF_MANY_CHILDREN |
```
Comment (EnTeQuAk) [inline docs/rules.md]: I would name this `RDF_TOO_MANY_CHILDREN` which is more explicit imho.

File: src/io/xpi.js [hunk 2]
```diff
@@ -140,7 +142,8 @@ export class Xpi extends IOBase {
 
   getChunkAsBuffer(path, chunkLength) {
     return new Promise((resolve, reject) => {
-      this.checkPath(path);
+      const pathErr = this._checkPath(path);
+      if (pathErr) return reject(pathErr);
```
Comment (EnTeQuAk) [inline src/io/xpi.js]: I think from a semantic point of view I'd still like to have `checkPath` raise an exception but use `try/except` here to catch the exception and then `reject` it properly. I also don't think we should rename `checkPath` to `_checkPath` but that's nothing to worry about too much.

File: src/messages/rdf.js [hunk 3]
```diff
@@ -71,6 +71,30 @@ export const RDF_TYPE_MISSING = {
   file: INSTALL_RDF,
 };
 
+export function rdfMultipleTags(tag) {
+  return {
+    code: `RDF_MULTIPLE_${tag.toUpperCase()}_TAGS`,
+    message: _(`Multiple <${tag}> elements found`),
+    description: _(oneLine`There should be only one tag <${tag}>`),
```
Comment (EnTeQuAk) [inline src/messages/rdf.js]: You don't need the `oneLine` helper here. It's only used to strip newlines and whitespaces if a string spans across multiple lines.

File: src/messages/rdf.js [hunk 4]
```diff
@@ -71,6 +71,30 @@ export const RDF_TYPE_MISSING = {
...
+    file: INSTALL_RDF,
+  };
+}
+
+export const RDF_TAG_NOT_FOUND = {
+  code: 'RDF_TAG_NOT_FOUND',
+  message: _(`RDF Node is not defined`),
+  description: _(oneLine`RDF tag should be defined`),
```
Comment (EnTeQuAk) [inline src/messages/rdf.js]: Same here with `oneLine`

File: src/messages/rdf.js [hunk 5]
```diff
@@ -71,6 +71,30 @@ export const RDF_TYPE_MISSING = {
...
+  file: INSTALL_RDF,
+};
+
+export const RDF_MANY_CHILDREN = {
+  code: 'RDF_MANY_CHILDREN',
+  message: _(`RDF node should only have a single descendant <Description>`),
+  description: _(oneLine`RDF tag should have only one child`),
```
Comment (EnTeQuAk) [inline src/messages/rdf.js]: Here `oneLine` isn't really needed.

File: src/parsers/installrdf.js [hunk 6]
```diff
@@ -15,12 +15,23 @@ export default class InstallRdfParser {
...
-      restartless: this._getIsBootstrapped(),
+    return new Promise((resolve, reject) => {
+      const guid = this._getGUID();
+      const name = this._getName();
+      const type = this._getAddonType();
+      const version = this._getVersion();
+      const restartless = this._getIsBootstrapped();
+      if (guid === null || name === null || version === null) {
+        return reject(new Error(oneLine`Cannot get metadata,
+                                  version, name, or id tag is missing`));
```
Comment (EnTeQuAk) [inline src/parsers/installrdf.js]: You can put the whole string on the next line to make it easier to read, something like:

```javascript
return reject(new Error(oneLine`
    Cannot get metadata, version, name or id tag is missing`));
```

Also, I'd write the error message a bit differently: `Cannot get metadata, version, name or id tag.` The `is missing` part is a bit confusing imho.

**Comment by EnTeQuAk** [comment]
Soooooo, we just had a discussion in IRC and I falsely marked #351 as a good first bug while we are actually pretty ready to remove all RDF related stuff from this code-base.

Languagepacks are moving into WebExtension direction right now, we never had dictionary validation in the linter and there isn't much else that needs RDF to work.

So, unfortunately I'm going to close this PR BUT as a compensation if you want, you can give https://github.com/mozilla/addons-linter/pull/1308 a look and maybe re-base or redo the whole removal part which should be pretty fun I hope.

Sorry for keeping you waiting so long and thanks a lot for all the work that you invested into this!

**Pull request closed without merging**